storm-core/src/jvm/org/apache/storm/cluster/ZKStateStorage.java

```diff
@@ -0,0 +1,270 @@
+
+import org.apache.curator.framework.CuratorFramework;
+import org.apache.curator.framework.state.*;
+import org.apache.curator.framework.state.ConnectionState;
+import org.apache.storm.Config;
+import org.apache.storm.callback.DefaultWatcherCallBack;
+import org.apache.storm.callback.WatcherCallBack;
+import org.apache.storm.callback.ZKStateChangedCallback;
+import org.apache.storm.utils.Utils;
+import org.apache.storm.zookeeper.Zookeeper;
+import org.apache.zookeeper.CreateMode;
+import org.apache.zookeeper.KeeperException;
+import org.apache.zookeeper.Watcher;
+import org.apache.zookeeper.data.ACL;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class ZKStateStorage implements IStateStorage {
+
+    private static Logger LOG = LoggerFactory.getLogger(ZKStateStorage.class);
+
+    private ConcurrentHashMap<String, ZKStateChangedCallback> callbacks = new ConcurrentHashMap<String, ZKStateChangedCallback>();
+    private CuratorFramework zkWriter;
+    private CuratorFramework zkReader;
+    private AtomicBoolean active;
+
+    private boolean isNimbus;
+    private Map authConf;
+    private Map<Object, Object> conf;
+
+    public ZKStateStorage(Map<Object, Object> conf, Map authConf, List<ACL> acls, ClusterStateContext context) throws Exception {
+        this.conf = conf;
+        this.authConf = authConf;
+        if (context.getDaemonType().equals(DaemonType.NIMBUS))
+            this.isNimbus = true;
+
+        // just mkdir STORM_ZOOKEEPER_ROOT dir
+        CuratorFramework zkTemp = mkZk();
+        String rootPath = String.valueOf(conf.get(Config.STORM_ZOOKEEPER_ROOT));
+        Zookeeper.mkdirs(zkTemp, rootPath, acls);
+        zkTemp.close();
+
+        active = new AtomicBoolean(true);
+        zkWriter = mkZk(new WatcherCallBack() {
+            @Override
+            public void execute(Watcher.Event.KeeperState state, Watcher.Event.EventType type, String path) {
+                if (active.get()) {
+                    if (!(state.equals(Watcher.Event.KeeperState.SyncConnected))) {
+                        LOG.warn("Received event {} : {}: {} with disconnected Zookeeper.", state, type, path);
+                    } else {
+                        LOG.info("Received event {} : {} : {}", state, type, path);
+                    }
+
+                    if (!type.equals(Watcher.Event.EventType.None)) {
+                        for (Map.Entry<String, ZKStateChangedCallback> e : callbacks.entrySet()) {
+                            ZKStateChangedCallback fn = e.getValue();
+                            fn.changed(type, path);
+                        }
+                    }
+                }
+            }
+        });
+        if (isNimbus) {
+            zkReader = mkZk(new WatcherCallBack() {
```

Comment:
It might be nice to reuse the same watcher callback instance, just so we don't have two identical anonymous classes.